doc/source/reference/arrays.promotion.rst

```diff
@@ -201,6 +201,25 @@ This leads to what may appear as "exceptions" to the rules:
 In principle, some of these exceptions may make sense for other functions.
 Please raise an issue if you feel this is the case.
 
+Notable behavior with Python builtin type classes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When combining Python's builtin scalar *types* (i.e., ``float``, ``int``,
+or ``complex``, not scalar *values*), the promotion rules can appear
+surprising:
+
+  >>> np.result_type(7, np.array([1], np.float32))
+  dtype('float32')  # The scalar value '7' does not impact type promotion
+  >>> np.result_type(type(7), np.array([1], np.float32))
+  dtype('float64')  # The *type* of the scalar value '7' does impact promotion
+  # Similar situations happen with Python's float and complex types
+
+The reason for this behavior is that NumPy converts ``int`` to its default
+integer type, and uses that type for promotion:
```

Comment:
Side note: I could not find good documentation on default integer type besides the somewhat confusing https://numpy.org/devdocs/numpy_2_0_migration_guide.html#windows-default-integer; perhaps we could a subsection somewhere in this file?

Reply:
Sorry, I was out while this happened. But would also value someone else's input here.

To me, this is a detail about `result_type`.  And I feel that it should be mentioned there mainly?
(It is also something that we probably should consider changing there!)

Reply:
The default types are documented in the list of types, I think.  But, I suppose we could move the footnote about it up and just state the fact that the default is `int64` on 64bit systems (which are most systems, so we use that).

Reply:
Thanks, I added a reference to the promotion list in 904a6f5. I also reworded the footnote slightly.

> To me, this is a detail about result_type. And I feel that it should be mentioned there mainly?
(It is also something that we probably should consider changing there!)

Do you mean, should this whole subsection be moved to the `result_type` doc, or just the default integer discussion?

Reply:
I was wondering about this whole addition.  The reason is that I don't think we mention the syntax of `result_type` to allow dtypes at all (which is the crux, `int` here would ideally not be a dtype instance, but an abstract DType class/type.  But `result_tyee` uses `np.dtype(int)` which does a silly guess at the default).

So I am wondering if adding it here just adds another layer of complexity that is mainly needed with `result_type`.  OTOH, explaining `result_type` in more detail here could make sense...

Reply:
I see. Just as an additional wrinkle, promote_types also accepts int...